src/libsync/propagatedownload.cpp

```diff
@@ -821,8 +827,26 @@ void PropagateDownloadFile::slotGetFinished()
     validator->start(_tmpFile.fileName(), checksumHeader);
 }
 
-void PropagateDownloadFile::slotChecksumFail(const QString &errMsg)
+void PropagateDownloadFile::slotChecksumFail(const QString &errMsg, const QByteArray &checksumType, const QByteArray &checksum, const QString &filePath)
 {
+    if (!checksumType.isEmpty() && !checksum.isEmpty() && !filePath.isEmpty()) {
+        ConfigFile cfgFile;
+
+        if (cfgFile.allowChecksumValidationFail()) {
+            const QString key = checksum + _item->_fileId;
+            const auto numChecksumMismatchEntries = propagator()->_journal->keyValueStoreGetInt(QString(key), 0);
+            if (numChecksumMismatchEntries < numChecksumFailuresAllowed) {
+                qCWarning(lcPropagateDownload) << "Checksum validation has failed" << numChecksumMismatchEntries << " times, but, allowChecksumValidationFail is set.Let's give it another try...";
+                propagator()->_journal->keyValueStoreSet(QString(key), numChecksumMismatchEntries + 1);
```

Comment:
we should not only store the retry counter but also the checksum value. We really want to detect that we always compute the same value before we accept to still download the file:

First scenario
1. first try checksum value is `abcd`, server value is `edfr`
2. second try checksum value is `abcd`, server value is `edfr`
3. we ignore the mismatch and use the downloaded file

Second scenario
1. first try checksum value is `abcd`, server value is `edfr`
2. second try checksum value is `dcba`, server value is `edfr`
3. we take the mismatch arror into account and block this file from being used

Reply:
@mgallien I am storing checksum value as a key.  Here `const QString key = checksum + _item->_fileId;` where `checksum` is a checksum. Or, am I missing something?

Reply:
if the second time you download a file and you get the expected checksum how do you clean up the wrong checksum entry ?
I would only use the file to identify the entry to be sure that one can do the clean up